c/csrc/btcspv.c

```diff
@@ -0,0 +1,481 @@
+  return seq;
+}
+
+uint32_t btcspv_extract_sequence_witness(const_view_t *tx_in) {
+  const_view_t seq = btcspv_extract_sequence_le_witness(tx_in);
+  return AS_LE_UINT32(seq.loc);
+}
+
+script_sig_t btcspv_extract_script_sig_len(const_view_t *tx_in) {
+  const uint8_t tag = (tx_in->loc)[36];
+
+  uint8_t script_sig_len = 0;
+  uint8_t data_len = btcspv_determine_var_int_data_length(tag);
+
+  for (uint8_t i = 0; i < data_len; i++) {
+    script_sig_len += (tx_in->loc)[37 + i] << (i * 8);
+  }
+
+  if (script_sig_len == 0 && tag < 0xfd) {
+    script_sig_len = tag;
+  }
+
+  script_sig_t res = {data_len, script_sig_len};
+  return res;
+}
+
+byte_view_t btcspv_extract_script_sig(const_view_t *tx_in) {
+  const script_sig_t ss = btcspv_extract_script_sig_len(tx_in);
+  uint32_t length = 1 + ss.var_int_len + ss.script_sig_len;
+
+  byte_view_t script_sig = {(tx_in->loc) + 36, length};
+  return script_sig;
+}
+
+byte_view_t btcspv_extract_sequence_le_legacy(const_view_t *tx_in) {
+  const script_sig_t ss = btcspv_extract_script_sig_len(tx_in);
+  const uint32_t offset = 36 + 1 + ss.var_int_len + ss.script_sig_len;
+
+  byte_view_t seq = {(tx_in->loc) + offset, 4};
+  return seq;
+}
+
+uint32_t btcspv_extract_sequence_legacy(const_view_t *tx_in) {
+  const_view_t seq = btcspv_extract_sequence_le_legacy(tx_in);
+  return AS_LE_UINT32(seq.loc);
+}
+
+uint32_t btcspv_determine_input_length(const_view_t *tx_in) {
+  const script_sig_t ss = btcspv_extract_script_sig_len(tx_in);
+  return 41 + ss.var_int_len + ss.script_sig_len;
+}
+
+byte_view_t btcspv_extract_input_at_index(const_view_t *vin, uint8_t index) {
+  uint32_t length = 0;
+  uint32_t offset = 1;
+
+  for (uint8_t i = 0; i < index + 1; i++) {
+    const_view_t remaining = {(vin->loc) + offset, (vin->len) - offset};
+    length = btcspv_determine_input_length(&remaining);
+    if (i != index) {
+      offset += length;
+    }
+  }
+
+  byte_view_t input = {(vin->loc) + offset, length};
+  return input;
+}
+
+byte_view_t btcspv_extract_outpoint(const_view_t *tx_in) {
+  byte_view_t outpoint = {tx_in->loc, 36};
+  return outpoint;
+}
+
+byte_view_t btcspv_extract_input_tx_id_le(const_view_t *tx_in) {
+  byte_view_t tx_id_le = {tx_in->loc, 32};
+  return tx_id_le;
+}
+
+byte_view_t btcspv_extract_tx_index_le(const_view_t *tx_in) {
+  byte_view_t idx = {tx_in->loc + 32, 4};
+  return idx;
+}
+
+uint32_t btcspv_extract_tx_index(const_view_t *tx_in) {
+  const_view_t idx = btcspv_extract_tx_index_le(tx_in);
+  return AS_LE_UINT32(idx.loc);
+}
+
+//
+// Output Functions
+//
+
+uint32_t btcspv_determine_output_length(const_view_t *tx_out) {
+  uint8_t script_len = tx_out->loc[8];
+  if (script_len >= 0xfd) {
+    return 0;
+  }
+  return script_len + 9;
+}
+
+byte_view_t btcspv_extract_output_at_index(const_view_t *vout, uint8_t index) {
+  byte_view_t *remaining = NULL;
+
+  uint32_t output_len = 0;
+  uint32_t offset = 1;
+  uint8_t idx = index;
+
+  for (int i = 0; i < index + 1; i++) {
```

Comment:
If `index` is `0xFF`, `index + 1` will be 0, which means this loop won't run. This is also why gcc generates that warning you are ignoring below.

Reply:
Ah thanks :+1:. To match behavior of other implementations, this should actually error when `index > 0xfc`, as the vin/vout parsers don't currently support multi-byte varints. Will add this check in the next commit